File: frontend/src/lib/contextUpdates.js
import { parseUtc } from "./dateUtils";

const SKIP = new Set(["", "n/a", "unknown", "0", "profiling in progress"]);

function tsMs(value) {
  const d = parseUtc(value);
  return d ? d.getTime() : null;
}

function displayDate(value) {
  const d = parseUtc(value);
  if (!d) return "—";
  return d.toLocaleDateString("en-GB", {
    timeZone: "Asia/Kolkata",
    day: "2-digit",
    month: "2-digit",
    year: "2-digit",
  });
}

function durationLabel(seconds) {
  const n = Math.floor(Number(seconds) || 0);
  if (n <= 0) return "";
  const mins = Math.floor(n / 60);
  const secs = n % 60;
  if (mins > 0) return ` (${mins}m ${secs}s)`;
  return ` (${secs}s)`;
}

function hasValue(value) {
  if (value == null) return false;
  const s = String(value).trim();
  return Boolean(s) && !SKIP.has(s.toLowerCase());
}

function entry({ at, icon, type, context, sortMs }) {
  const ms = sortMs != null ? sortMs : tsMs(at);
  return {
    at: at || "",
    date: displayDate(at),
    icon,
    type,
    context,
    sort_ms: ms || 0,
  };
}

/** Client fallback when API has not backfilled context_updates yet. */
export function buildContextUpdatesFromLeadAndCalls(lead, calls = []) {
  const entries = [];
  const rows = [...(calls || [])].sort(
    (a, b) =>
      (tsMs(b.created_at || b.call_date) || 0) - (tsMs(a.created_at || a.call_date) || 0)
  );

  let earliestMs = null;
  for (const call of rows) {
    const rawTs = call.created_at || call.call_date;
    const ms = tsMs(rawTs);
    if (ms != null && (earliestMs == null || ms < earliestMs)) earliestMs = ms;

    const label = (call.disposition || call.status || "completed").trim();
    entries.push(
      entry({
        at: rawTs,
        icon: "phone",
        type: "call",
        context: `Outbound call — ${label}${durationLabel(call.duration)}`,
        sortMs: ms,
      })
    );
  }

  if (hasValue(lead?.budget)) {
    entries.push(
      entry({
        at: lead.updated_at || lead.last_call_date,
        icon: "whatsapp",
        type: "whatsapp",
        context: `Budget: ${lead.budget}`,
      })
    );
  }

  if (hasValue(lead?.configuration)) {
    entries.push(
      entry({
        at: lead.updated_at || lead.last_call_date,
        icon: "phone",
        type: "call",
        context: `Interested in ${lead.configuration} configuration`,
      })
    );
  }

  if (hasValue(lead?.project)) {
    entries.push(
      entry({
        at: lead.updated_at || lead.last_call_date,
        icon: "human",
        type: "human",
        context: `Showed interest in ${lead.project}`,
      })
    );
  }

  if (earliestMs != null) {
    entries.push(
      entry({
        at: new Date(earliestMs).toISOString(),
        icon: "phone",
        type: "call",
        context: "Initial contact made",
        sortMs: earliestMs,
      })
    );
  }

  if (!entries.length) {
    return [entry({ at: null, icon: "phone", type: "call", context: "Initial contact made", sortMs: 0 })];
  }

  entries.sort((a, b) => (b.sort_ms || 0) - (a.sort_ms || 0));
  const seen = new Set();
  return entries
    .filter((e) => {
      const key = `${e.context}|${e.sort_ms}`;
      if (seen.has(key)) return false;
      seen.add(key);
      return true;
    })
    .map(({ sort_ms, ...rest }) => rest);
}

File: backend/app/utils/context_updates.py
"""Build and persist lead context_updates from calls and CRM fields."""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from ..core.time_utils import serialize_datetime_utc
from .lead_call_history import build_lead_call_history_query

logger = logging.getLogger(__name__)

_SKIP_VALUES = frozenset({"", "n/a", "unknown", "0", "profiling in progress"})


def _ts_ms(value: Any) -> Optional[int]:
    if value is None or value == "":
        return None
    if isinstance(value, datetime):
        dt = value
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return int(dt.timestamp() * 1000)
    s = str(value).strip()
    if not s:
        return None
    try:
        normalized = s if s.endswith("Z") or "+" in s[-6:] else (f"{s}Z" if "T" in s else f"{s}T00:00:00Z")
        dt = datetime.fromisoformat(normalized.replace("Z", "+00:00"))
        return int(dt.timestamp() * 1000)
    except (TypeError, ValueError):
        return None


def _display_date(value: Any) -> str:
    ms = _ts_ms(value)
    if ms is None:
        return "—"
    dt = datetime.fromtimestamp(ms / 1000, tz=timezone.utc)
    return dt.strftime("%d/%m/%y")


def _has_meaningful(value: Any) -> bool:
    if value is None:
        return False
    s = str(value).strip()
    return bool(s) and s.lower() not in _SKIP_VALUES


def _duration_label(seconds: int) -> str:
    n = int(seconds or 0)
    if n <= 0:
        return ""
    mins, secs = divmod(n, 60)
    if mins > 0:
        return f" ({mins}m {secs}s)"
    return f" ({secs}s)"


def _entry(
    *,
    at: Any,
    icon: str,
    type_: str,
    context: str,
    sort_ms: Optional[int] = None,
) -> Dict[str, Any]:
    ms = sort_ms if sort_ms is not None else _ts_ms(at)
    return {
        "at": serialize_datetime_utc(at) if at else "",
        "date": _display_date(at),
        "icon": icon,
        "type": type_,
        "context": context,
        "sort_ms": ms or 0,
    }


def build_context_updates(lead: Dict[str, Any], calls: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Build timeline entries (newest first)."""
    entries: List[Dict[str, Any]] = []
    lead = lead or {}

    call_rows = sorted(
        calls or [],
        key=lambda c: _ts_ms(c.get("created_at") or c.get("call_date")) or 0,
        reverse=True,
    )

    earliest_ms: Optional[int] = None
    for call in call_rows:
        raw_ts = call.get("created_at") or call.get("call_date")
        ms = _ts_ms(raw_ts)
        if ms is not None and (earliest_ms is None or ms < earliest_ms):
            earliest_ms = ms

        label = (call.get("disposition") or call.get("status") or "completed").strip()
        dur = int(call.get("duration", 0) or 0)
        entries.append(
            _entry(
                at=raw_ts,
                icon="phone",
                type_="call",
                context=f"Outbound call — {label}{_duration_label(dur)}",
                sort_ms=ms,
            )
        )

        se = call.get("structured_extraction")
        if isinstance(se, dict):
            summary = (se.get("call_summary") or "").strip()
            if summary:
                entries.append(
                    _entry(
                        at=raw_ts,
                        icon="human",
                        type_="human",
                        context=summary[:280],
                        sort_ms=ms,
                    )
                )

    if _has_meaningful(lead.get("sales_qualification")):
        at = lead.get("sales_qualified_at") or lead.get("updated_at")
        entries.append(
            _entry(
                at=at,
                icon="human",
                type_="human",
                context=f"Sales qualification: {lead.get('sales_qualification')}",
            )
        )

    if _has_meaningful(lead.get("budget")):
        at = lead.get("updated_at") or lead.get("last_call_date")
        entries.append(
            _entry(
                at=at,
                icon="whatsapp",
                type_="whatsapp",
                context=f"Budget: {lead.get('budget')}",
            )
        )

    if _has_meaningful(lead.get("configuration")):
        at = lead.get("updated_at") or lead.get("last_call_date")
        entries.append(
            _entry(
                at=at,
                icon="phone",
                type_="call",
                context=f"Interested in {lead.get('configuration')} configuration",
            )
        )

    if _has_meaningful(lead.get("project")):
        at = lead.get("updated_at") or lead.get("last_call_date")
        entries.append(
            _entry(
                at=at,
                icon="human",
                type_="human",
                context=f"Showed interest in {lead.get('project')}",
            )
        )

    if earliest_ms is not None:
        entries.append(
            _entry(
                at=datetime.fromtimestamp(earliest_ms / 1000, tz=timezone.utc),
                icon="phone",
                type_="call",
                context="Initial contact made",
                sort_ms=earliest_ms,
            )
        )

    if not entries:
        return [
            _entry(
                at=None,
                icon="phone",
                type_="call",
                context="Initial contact made",
                sort_ms=0,
            )
        ]

    entries.sort(key=lambda e: e.get("sort_ms", 0), reverse=True)
    seen_context: set = set()
    deduped: List[Dict[str, Any]] = []
    for e in entries:
        key = (e.get("context"), e.get("sort_ms"))
        if key in seen_context:
            continue
        seen_context.add(key)
        out = {k: v for k, v in e.items() if k != "sort_ms"}
        deduped.append(out)
    return deduped


async def fetch_calls_for_lead(db, lead_id: str) -> List[Dict[str, Any]]:
    """Load call_history rows for a lead (same logic as GET /leads/{id}/calls)."""
    lead = await db.leads.find_one({"id": lead_id}, {"_id": 0})
    if not lead:
        return []

    mobile_digits = (lead.get("mobile_digits") or "").strip()
    if not mobile_digits and not lead.get("id"):
        return []

    history_docs = await db.call_history.find(
        build_lead_call_history_query(lead_id, lead),
        {"_id": 0},
    ).sort("created_at", -1).to_list(50)

    calls: List[Dict[str, Any]] = []
    for d in history_docs:
        ts = d.get("started_at") or d.get("created_at")
        se = d.get("structured_extraction") or {}
        calls.append(
            {
                "created_at": serialize_datetime_utc(ts),
                "call_date": serialize_datetime_utc(ts),
                "status": d.get("status", ""),
                "disposition": d.get("disposition", ""),
                "duration": int(d.get("duration", 0) or 0),
                "structured_extraction": se if isinstance(se, dict) else {},
            }
        )
    return calls


async def persist_lead_context_updates(db, lead_id: str) -> None:
    """Recompute and store context_updates on the lead document."""
    lead = await db.leads.find_one({"id": lead_id}, {"_id": 0})
    if not lead:
        return
    calls = await fetch_calls_for_lead(db, lead_id)
    updates = build_context_updates(lead, calls)
    await db.leads.update_one(
        {"id": lead_id},
        {"$set": {"context_updates": updates, "updated_at": datetime.now(timezone.utc)}},
    )
